gr-blocks/python/blocks/qa_rotator_cc.py

```diff
@@ -0,0 +1,199 @@
+# any later version.
+#
+# GNU Radio is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with GNU Radio; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+
+from gnuradio import gr, gr_unittest
+from gnuradio import blocks
+import numpy as np
+import pmt
+import time
+
+
+class qa_rotator_cc(gr_unittest.TestCase):
+
+    def setUp(self):
+        self.n_samples  = 100
+        self.f_in       = 0.01  # frequency on rotator's input
+        self.f_shift    = 0.015 # rotator's starting frequency
+        tag_inc_updates = True
+
+        self.tb = gr.top_block()
+
+        # Input IQ samples
+        in_angles  = 2 * np.pi * np.arange(self.n_samples) * self.f_in
+        in_samples = np.exp(1j*in_angles)
+
+        # Rotator's starting phase increment
+        phase_inc = 2 * np.pi * self.f_shift
+
+        self.source     = blocks.vector_source_c(in_samples)
+        self.rotator_cc = blocks.rotator_cc(phase_inc, tag_inc_updates)
+        self.sink       = blocks.vector_sink_c()
+
+    def tearDown(self):
+        self.tb = None
+
+    def test_001_t(self):
+        """Complex sinusoid frequency shift
+        """
+
+        f_out = self.f_in + self.f_shift # frequency of rotator's output
+
+        # Expected IQ samples
+        expected_angles  = 2 * np.pi * np.arange(self.n_samples) * f_out
+        expected_samples = np.exp(1j*expected_angles)
+
+        self.tb.connect(self.source, self.rotator_cc, self.sink)
+        self.tb.run()
+
+        for (x,y) in zip(self.sink.data(), expected_samples):
+            self.assertComplexAlmostEqual(x, y, places=5)
+
+    def test_002_t(self):
+        """Phase increment update at specified offset via control message
+        """
+
+        new_f_shift   = 0.016  # new rotator's frequency
+        offset        = 50     # when to apply new increment
+        msg_period_ms = 0      # periodicity (in ms) of increment update message
+
+        new_phase_inc = float(2 * np.pi * new_f_shift)
+        f_out_1       = self.f_in + self.f_shift # before update
+        f_out_2       = self.f_in + new_f_shift  # after update
+
+        # Message to be sent to the rotator in order to update its increment
+        ctrl_msg = pmt.make_dict()
+        ctrl_msg = pmt.dict_add(ctrl_msg,
+                                pmt.intern("inc"),
+                                pmt.from_double(new_phase_inc))
+        ctrl_msg = pmt.dict_add(ctrl_msg,
+                                pmt.intern("offset"),
+                                pmt.from_uint64(offset))
+
+        rot_ctrl = blocks.message_strobe(ctrl_msg, msg_period_ms)
```

Comment:
This isn't very reliable and I accept suggestions. I'm using message period as 0 so that I try to get the message sent before processing the 100 samples that are tested. However, this does not work every time. Sometimes the samples are processed first. Hints?

Reply:
@mbr0wn aside from this, I believe this MR is ready. If you have suggestions on how to solve the above testing issue, let me know.

To clarify, the goal here is to test a message sent to the rotator asking for an update of its's phase  increment. To do so, the message has to be sent first, before the rotator actually processes the samples. In this particular test the message asks for an update on sample offset 50, so `blocks.message_strobe` should convey the message before `blocks.rotator_cc` processes the first 50 samples, otherwise the phase increment will not be updated. This does not seem to work reliably using `blocks.message_strobe`, so not always the test will pass.

Reply:
If you call `to_basic_block()._post()` to pass in the message instead of using a message strobe, you might be able to force the message into the queue before you start the FG.

Reply:
Quick grep found this: https://github.com/gnuradio/gnuradio/blob/64ee803068269fe3b9563fd415c420fb513a1c9b/gr-digital/python/digital/qa_header_payload_demux.py#L332-L336

This might work for you also.

Reply:
Thanks. I used `insert_tail`, which is what would `_post` would call AFAIU.